packages/apollo-server-core/src/caching.ts

```diff
@@ -1,4 +1,67 @@
+import { ExecutionResult } from 'graphql';
+
 export interface PersistedQueryCache {
   set(key: string, data: string): Promise<any>;
   get(key: string): Promise<string | null>;
 }
+
+export function calculateCacheControlHeaders(
+  responses: Array<ExecutionResult & { extensions?: Record<string, any> }>,
+) {
+  let maxAge = Number.MAX_VALUE;
+  let publicOrPrivate = 'public';
+
+  //Because of the early exit, we are unable to use forEach. While a reduce
+  //loop might be possible, a for loop is more readable
+  for (let i = 0; i < responses.length; i++) {
+    const response = responses[i];
+
+    const cacheControl: {
+      version: number;
+      hints: Array<{ scope?: string; maxAge?: number; path: Array<string> }>;
+    } =
+      response.extensions && response.extensions.cacheControl;
+
+    //If there are no extensions or hints, then the headers should not be present
+    if (
+      !cacheControl ||
+      !cacheControl.hints ||
+      cacheControl.hints.length === 0
+    ) {
+      return {};
+    }
+
+    const rootHints = new Set<string>();
+    for (let y = 0; y < cacheControl.hints.length; y++) {
```

Comment:
grab cacheControl.hints[y] immediately like you did in the outer loop?  (Also a little odd that the loop indices are i/y.)